[Tan Ming Li, Natalie] iP

NatalieTanML opened the pull request:
### Jarvis

> "J.A.R.V.I.S., are you up?"
> "For you sir, always."

Jarvis is your personal _(task)_ assistant. He's
- text-based,
- easy to learn, and
- ~fast~ **SUPER** fast to use!

To get started with Jarvis,
1. Download Jarvis from [this link](https://github.com/NatalieTanML/ip/releases/download/A-Jar/Jarvis.jar).
2. Launch the app.
3. Go crazy 🤪

Jarvis can do the following for you:
- [x] Manage your tasks, deadlines and events
- [x] Provide an overview of what's in your task list
- [ ] Give reminders (coming soon!)

If you're a Java programmer, you can use it to practice Java too. Here's a snippet of the `startup` function:
```java
public static void startup() throws JarvisException {
    storage = new Storage("data/data.txt");
    ui = new Ui();
    tasks = new TaskList(storage.loadData(), ui);
    ui.welcome();
}
```

Commit:
Add Gradle support

NatalieTanML committed:
Implemented Level-1

NatalieTanML committed:
Implemented Level-2

NatalieTanML committed:
Added Task class

NatalieTanML committed:
Implemented marking tasks as done/undone

NatalieTanML committed:
Abstracted code out of switch into methods

NatalieTanML committed:
Added inheritence to support multiple task types

Implemented Todo, Deadline and Event classes

Updated Task to be abstract

NatalieTanML committed:
Implemented support for tracking diff. task types

Implemented Todos, Deadlines and Events in Duke

Formatted/cleaned up code

NatalieTanML committed:
Implemented automated Text UI testing

NatalieTanML committed:
Added DukeException class

Moved Task-related classes into task package

NatalieTanML committed:
Import task packages

Set packages

NatalieTanML committed:
Implemented error handling

NatalieTanML committed:
Implemented support for deleting tasks

NatalieTanML committed:
Added input command enums

NatalieTanML committed:
Update indentation for switch cases

NatalieTanML committed:
Implement local data saving and retrieval

NatalieTanML committed:
Implement datetime parsing of deadlines and events

Currently, Jarvis does not do any validation for the dates in
deadline and event tasks. This allows users to enter non-date inputs
as the date for such tasks.

By providing datetime parsing, this allows Jarvis to do validation on
the input and could be extended into other features in the future.

NatalieTanML committed:
Merge branch 'branch-Level-7'

NatalieTanML committed:
Merge branch 'master' into branch-Level-8

NatalieTanML committed:
Rename program to Jarvis

NatalieTanML committed:
Abstract UI functionality into Ui class

NatalieTanML committed:
Abstract Task related methods into TaskList class

NatalieTanML committed:
Abstract data storage handling into Storage class

NatalieTanML committed:
Abstract input parsing functions into Parser class

NatalieTanML committed:
Implement JUnit tests for Parser

NatalieTanML committed:
Update gitignore to ignore manifest

NatalieTanML committed:
Add JavaDocs to methods and classes

NatalieTanML committed:
Implement find command that allows task searching

NatalieTanML committed:
Clean up code to comply with coding standards

NatalieTanML committed:
Merge branch 'refs/heads/A-JavaDoc'

NatalieTanML committed:
Merge branch 'A-CodingStandard'

NatalieTanML committed:
Merge branch 'Level-9'

NatalieTanML committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

NatalieTanML committed:
Add gradle

NatalieTanML committed:
Add Checkstyle

NatalieTanML committed:
Refactor code to adhere to checkstyle

NatalieTanML committed:
Implement first iteration of GUI for Jarvis

NatalieTanML committed:
Implement working GUI with Jarvis

NatalieTanML committed:
Edit code to adhere to checkstyle

pyk595 commented on src/main/java/enums/Command.java:
```diff
@@ -0,0 +1,18 @@
+package enums;
+
+public enum Command {
```
You can consider writing JavaDoc comments for this enum class? Although it is a reach goal since you only need to do for half of the non-private classes and methods.

pyk595 commented on src/main/java/enums/Command.java:
```diff
@@ -0,0 +1,18 @@
+    EVENT,
+    FIND;
+
+    @Override
```
Same as above for overridden methods, since you are changing the behaviour of the method.

pyk595 commented on src/main/java/exception/JarvisException.java:
```diff
@@ -0,0 +1,11 @@
+        super();
+    }
+
+    public JarvisException(String message) {
```
Very clean and good use of inheritance, but might confuse other developers on what error this exception is supposed to handle/catch so maybe JavaDoc comments might help?

pyk595 commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,90 @@
+    private TaskList tasks;
+    private Storage storage;
+
+    private boolean processNext = true;
```
Maybe you can consider changing processNext to a question like isProcessNext to follow coding standard?

jinnhl replied:
Should the variable processNext have a more boolean name? Maybe hasNext?

pyk595 commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    /**
+     * Saves the program state into local storage and closes open resources.
```
Good JavaDocs for methods in this class! Follows coding standard! Maybe consider writing on more JavaDoc comment for the class to fully document this class?

pyk595 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,104 @@
+                throw new JarvisException("The keyword cannot be empty.");
+            }
+        }
+
```
Can consider deleting this line because unnecessary whitespace?

pyk595 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public String toFileString() {
```
Same thing as above, can consider writing JavaDoc comment for your overridden methods for more details on the behaviour?

pyk595 reviewed (commented):
Ignore this comment sorry!

pyk595 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,104 @@
+        parsedCommand.put("command", tokens[0].trim().toUpperCase());
+
+        Command command = Command.valueOf((String) parsedCommand.get("command"));
+        if (command == Command.MARK
```
A lot of if-elses here, might be a bit messy. You can consider creating specific exceptions to be thrown, and use a switch case and a single try-catch to handle all the exceptions that could be thrown.

pyk595 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,94 @@
+                Scanner s = new Scanner(dataFile);
+                while (s.hasNext()) {
+                    String[] line = s.nextLine().split("\\|");
+                    Task task;
```
Maybe you might want to declare the task variable inside the switch cases? 
Something like `Task task = new Todo(line[2]);`

pyk595 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     * @return List of tasks stored in the file.
+     * @throws JarvisException If file retrieval failed.
+     */
+    public List<Task> loadData() throws JarvisException {
```
Good implementation of "Make the happy path prominent", you put the expected behaviour first, and subsequently the potential problem path of directory not existing next.

pyk595 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,37 @@
+        return (isDone ? "[X]" : "[ ]"); // mark done task with X
+    }
+
+    public void markAsDone() {
```
Good to see that you are using verbs for your methods and nouns for your tasks

pyk595 reviewed (commented):
Good job with code quality, I think your code is generally clear, just that your if-else chunk can potentially be a point of error so I suggested a possible alternative!

jinnhl commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    public String getResponse(String input) {
+        while (processNext) {
+            try {
```
Should parsedCommand be renamed to parsedCommands since it is a HashMap, which is itself a collection?

jinnhl commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    public String markAsDone(HashMap<String, Object> parsedCommand) throws JarvisException {
+        checkIfEmpty();
+        try {
```
Would it be easier if the singular task and the list tasks have a more distinct name from each other? That way it can be identified with one glance.

jinnhl reviewed (commented):
Ignore this review...

jinnhl commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,90 @@
+                switch (Command.valueOf((String) parsedCommand.get("command"))) {
+                case BYE:
+                    return shutdown();
+                case LIST:
```
I like the indentations for switch case, normally people would have case indented one above switch

jinnhl commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,90 @@
+    public String getResponse(String input) {
+        while (processNext) {
+            try {
+                HashMap<String, Object> parsedCommand = Parser.parseInput(input);
```
Should parsedCommand be renamed to parsedCommands since it is a HashMap, which is itself a collection?

jinnhl commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,176 @@
+    public String markAsDone(HashMap<String, Object> parsedCommand) throws JarvisException {
+        checkIfEmpty();
+        try {
+            Task task = tasks.get((Integer) parsedCommand.get("num"));
```
Would it be easier if the singular task and the list tasks have a more distinct name from each other? That way it can be identified with one glance.

jinnhl reviewed (commented):
Coding standards are adhered to very closely, just some very minor adjustments and it would be perfect! Good Job!

NatalieTanML committed:
Add assertions to DialogBox, MainWindow, Storage

Some of these methods use variables (either private or local) that are
assumed to be not null.

Currently, there are no additional checks done to ensure that these variables
are valid apart from general try-catch clauses.

Using assert statements ensure that the code will not run into exceptions
during runtime.

NatalieTanML committed:
Improve code quality and add comments

Certain variable names such as `processNext` and `tasks` were not
compliant with the recommended coding standard. Some methods and classes
did not have the proper Javadoc header comments as well.

By renaming those variables to `hasNext` and `taskList`, it follows the
Java coding standard more closely. Adding comments to the Enum class and
other methods gives other users/readers a better understanding on what
the code does as well.

NatalieTanML committed:
Merge pull request #1 from NatalieTanML/branch-A-Assertions

Add assertions to DialogBox, MainWindow, Storage

NatalieTanML committed:
Merge branch 'master' into branch-A-CodeQuality

NatalieTanML committed:
Merge pull request #2 from NatalieTanML/branch-A-CodeQuality

Improve code quality and add comments

NatalieTanML committed:
Add gradle.yml workflow file

NatalieTanML committed:
Implement snooze command for deadlines and events

NatalieTanML committed:
Implement rounded images and proper shutdown

NatalieTanML committed:
Implement user guide

NatalieTanML committed:
Set theme jekyll-theme-cayman

NatalieTanML committed:
Update README.md

NatalieTanML committed:
Fix typo in README

NatalieTanML committed:
Merge branch 'master' of https://github.com/NatalieTanML/ip
